### PR TITLE
[tests-only][full-ci] Add tests for zero-byte file upload on public share

### PR DESCRIPTION
**Pull request by nabim777**
## Description

In this PR, there is an added bug coverage test. Following scenarios are added for upload via webdav api:

```feature
Scenario: public uploads a zero byte file to a public share folder
Scenario: public uploads a zero byte file to a public share folder inside project space
Scenario: public uploads a zero byte file to a public share project space
```

## Related Issue

- https://github.com/owncloud/ocis/issues/10649
- Bug: https://github.com/owncloud/ocis/issues/10469
## How Has This Been Tested?

- test environment:
- CI
- locally
- ...

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

File: tests/acceptance/features/apiSpacesShares/shareUploadTUS.feature [hunk 1]
```diff
@@ -402,3 +402,50 @@ Feature: upload resources on share using TUS protocol
       | Tus-Resumable   | 1.0.0                     |
     Then the HTTP status code should be "460"
     And for user "Alice" the content of the file "/textFile.txt" of the space "Personal" should be "original content"
+
+  @issue-10649
+  Scenario: public uploads a zero byte file to a public share folder
```
Comment (nabim777) [inline tests/acceptance/features/apiSpacesShares/shareUploadTUS.feature:407]:
```suggestion
  Scenario: public uploads a zero byte file to a public share folder
```

Since existing test uses the TUS library and is unable to get and check the response.
So, there is no check for the `location` header on TUS upload.

File: tests/acceptance/bootstrap/TUSContext.php [hunk 2]
```diff
@@ -332,6 +332,43 @@ public function uploadFileUsingTus(
...
+	 * @param string $source
+	 * @param string $destination
+	 * @param string|null $password
+	 *
+	 * @return void
+	 */
+	public function publicUploadFileUsingTus(
+		string  $source,
+		string  $destination,
+		?string $password,
```
Comment (S-Panta) [inline tests/acceptance/bootstrap/TUSContext.php:345]: Why is password optional? I see no feature file with no password

File: tests/acceptance/features/apiSpaces/uploadSpaces.feature [hunk 3]
```diff
@@ -93,3 +93,49 @@ Feature: Upload files into a space
...
+    And for user "Alice" folder "uploadFolder" of the space "Personal" should not contain these files:
+      | textfile (1).txt |
+
+  @issue-10649
+  Scenario: public uploads a zero byte file to a public share folder inside project space
+    Given using SharingNG
+    And user "Alice" has created a folder "/uploadFolder" in space "Project Ceres"
+    And user "Alice" has created the following resource link share:
+      | resource        | uploadFolder |
+      | space           | Project Ceres     |
```
Comment (S-Panta) [inline tests/acceptance/features/apiSpaces/uploadSpaces.feature:119]:
```suggestion
      | resource        | uploadFolder |
      | space           | Project Ceres  |
```

File: tests/acceptance/features/apiSpaces/uploadSpaces.feature [hunk 4]
```diff
@@ -93,3 +93,49 @@ Feature: Upload files into a space
...
+    And for user "Alice" folder "uploadFolder" of the space "Project Ceres" should not contain these files:
+      | textfile (1).txt |
+
+  @issue-10649
+  Scenario: public uploads a zero byte file to a public share project space
+    Given using SharingNG
+    And user "Alice" has created the following space link share:
+      | space           | Project Ceres  |
+      | permissionsRole | createOnly     |
+      | password        | %public% |
```
Comment (S-Panta) [inline tests/acceptance/features/apiSpaces/uploadSpaces.feature:135]: Format the feature file

File: tests/acceptance/features/apiSpacesShares/shareUploadTUS.feature [hunk 5]
```diff
@@ -402,3 +402,50 @@ Feature: upload resources on share using TUS protocol
...
+  Scenario: public uploads a zero-byte file to a shared folder inside project space
+    Given using SharingNG
+    And the administrator has assigned the role "Space Admin" to user "Alice" using the Graph API
+    And user "Alice" has created a space "Project" with the default quota using the Graph API
+    And user "Alice" has created a folder "/uploadFolder" in space "Project"
+    And user "Alice" has created the following resource link share:
+      | resource        | uploadFolder |
+      | space           | Project     |
+      | permissionsRole | createOnly   |
+      | password        | %public%     |
```
Comment (S-Panta) [inline tests/acceptance/features/apiSpacesShares/shareUploadTUS.feature:431]:
```suggestion
      | resource        | uploadFolder |
      | space           | Project      |
      | permissionsRole | createOnly   |
      | password        | %public%     |
```

File: tests/acceptance/bootstrap/TUSContext.php [hunk 6]
```diff
@@ -332,6 +332,43 @@ public function uploadFileUsingTus(
 		}
 	}
 
+	/**
+	 * @param string $source
+	 * @param string $destination
+	 * @param string|null $password
```
Comment (PrajwolAmatya) [inline tests/acceptance/bootstrap/TUSContext.php:338]:
```suggestion
	 * @param string $password
```

**Comment by nabim777** [comment]
https://github.com/owncloud/ocis/blob/96f1a27d3e1cc004a57dced1d398c72af94d485e/tests/acceptance/features/apiSpacesShares/shareUploadTUS.feature#L407

Since existing test uses the TUS library and is unable to get and check the response.
So, there is no check for the location header on TUS upload.

File: tests/acceptance/bootstrap/TUSContext.php [hunk 7]
```diff
@@ -332,6 +332,43 @@ public function uploadFileUsingTus(
...
+		string  $source,
+		string  $destination,
+		string $password,
+	):void {
+		$password = $this->featureContext->getActualPassword($password);
+		if ($this->featureContext->isUsingSharingNG()) {
+			$token = $this->featureContext->shareNgGetLastCreatedLinkShareToken();
+		} else {
+			$token = $this->featureContext->getLastCreatedPublicShareToken();
+		}
```
Comment (PrajwolAmatya) [inline tests/acceptance/bootstrap/TUSContext.php:352]:
```suggestion
		$token = $this->featureContext->isUsingSharingNG() ? $this->featureContext->shareNgGetLastCreatedLinkShareToken() : $this->featureContext->getLastCreatedPublicShareToken()
```

File: tests/acceptance/features/apiSpacesShares/shareUploadTUS.feature [hunk 8]
```diff
@@ -402,3 +402,50 @@ Feature: upload resources on share using TUS protocol
...
+  @issue-10331 @issue-10469
+  Scenario: public uploads a zero byte file to a public share folder
+    Given using SharingNG
+    And user "Alice" has created folder "/uploadFolder"
+    And user "Alice" has created the following resource link share:
+      | resource        | uploadFolder |
+      | space           | Personal     |
+      | permissionsRole | createOnly   |
+      | password        | %public%     |
+    When the public uploads a file from "filesForUpload/zerobyte.txt" to "textfile.txt" via TUS inside last link shared folder with password "%public%" using the WebDAV API
```
Comment (anon-pradip) [inline tests/acceptance/features/apiSpacesShares/shareUploadTUS.feature:415]:
```suggestion
    When the public uploads file "filesForUpload/zerobyte.txt" to "textfile.txt" via TUS inside last link shared folder with password "%public%" using the WebDAV API
```

Comment (anon-pradip) [reply tests/acceptance/features/apiSpacesShares/shareUploadTUS.feature:415]: Just to match with other similar steps :slightly_smiling_face:

File: tests/acceptance/bootstrap/TUSContext.php (hunk 7 above)
Comment (nabim777) [reply tests/acceptance/bootstrap/TUSContext.php:352]: I think the current structure is suitable for this case, as it’s easier to read and debug. I’ll consider this approch for simple cases.

File: tests/acceptance/bootstrap/SpacesTUSContext.php [hunk 9]
```diff
@@ -90,6 +90,25 @@ public function userUploadsAFileViaTusInsideOfTheSpaceUsingTheWebdavApi(
...
+	 * @When the public uploads file :source to :destination via TUS inside last link shared folder with password :password using the WebDAV API
+	 *
+	 * @param string $source
+	 * @param string $destination
+	 * @param string $password
+	 *
+	 * @return void
+	 * @throws Exception
+	 */
+	public function publicUploadsAFileViaTusInsideOfTheSpaceUsingTheWebdavApi(
```
Comment (S-Panta) [inline tests/acceptance/bootstrap/SpacesTUSContext.php:103]:
```suggestion
	public function thePublicUploadsFileViaTusInsideLastSharedFolderWithPasswordUsingTheWebdavApi(
```
Make sure the function name matches step definition.

File: tests/acceptance/features/apiSpacesShares/shareUploadTUS.feature [hunk 10]
```diff
@@ -402,3 +402,50 @@ Feature: upload resources on share using TUS protocol
...
+  @issue-10331 @issue-10469
+  Scenario: public uploads a zero byte file to a public share folder
+    Given using SharingNG
+    And user "Alice" has created folder "/uploadFolder"
+    And user "Alice" has created the following resource link share:
+      | resource        | uploadFolder |
+      | space           | Personal     |
+      | permissionsRole | createOnly   |
+      | password        | %public%     |
+    When the public uploads file "filesForUpload/zerobyte.txt" to "textfile.txt" via TUS inside last link shared folder with password "%public%" using the WebDAV API
```
[the remark this answers is not in the document]
Comment (nabim777) [reply tests/acceptance/features/apiSpacesShares/shareUploadTUS.feature:415]: I think according to comment there was an issue on uploading zero bytes where `location` header was missing and that was fixed by this PR https://github.com/owncloud/ocis/pull/10500 .

**Comment by anon-pradip** [comment]
Seems okay for me :+1:

File: tests/acceptance/features/apiSpaces/uploadSpaces.feature [hunk 11]
```diff
@@ -93,3 +93,49 @@ Feature: Upload files into a space
...
+      | resource        | uploadFolder  |
+      | space           | Project Ceres |
+      | permissionsRole | createOnly    |
+      | password        | %public%      |
+    When the public uploads file "filesForUpload/zerobyte.txt" to "textfile.txt" inside last link shared folder with password "%public%" using the public WebDAV API
+    Then the HTTP status code should be "201"
+    And for user "Alice" folder "uploadFolder" of the space "Project Ceres" should contain these files:
+      | textfile.txt |
+    And for user "Alice" folder "uploadFolder" of the space "Project Ceres" should not contain these files:
+      | textfile (1).txt |
```
Comment (saw-jan) [inline tests/acceptance/features/apiSpaces/uploadSpaces.feature:127]: also add second file check
```suggestion
    And for user "Alice" folder "uploadFolder" of the space "Project Ceres" should not contain these files:
      | textfile (1).txt |
      | textfile (2).txt |
```

File: tests/acceptance/features/apiSpacesShares/shareUploadTUS.feature [hunk 12]
```diff
@@ -402,3 +402,50 @@ Feature: upload resources on share using TUS protocol
...
+    Given using SharingNG
+    And the administrator has assigned the role "Space Admin" to user "Alice" using the Graph API
+    And user "Alice" has created a space "Project" with the default quota using the Graph API
+    And user "Alice" has created a folder "/uploadFolder" in space "Project"
+    And user "Alice" has created the following resource link share:
+      | resource        | uploadFolder |
+      | space           | Project      |
+      | permissionsRole | createOnly   |
+      | password        | %public%     |
+    When the public uploads file "filesForUpload/zerobyte.txt" to "textfile.txt" via TUS inside last link shared folder with password "%public%" using the WebDAV API
```
Comment (saw-jan) [inline tests/acceptance/features/apiSpacesShares/shareUploadTUS.feature:432]: need to check the `Location` header. Please, do in this PR

**Review by saw-jan — CHANGES_REQUESTED** [review]

File: tests/acceptance/features/apiSpacesShares/shareUploadTUS.feature (hunk 12 above)
Comment (nabim777) [reply tests/acceptance/features/apiSpacesShares/shareUploadTUS.feature]: I have updated this PR. I have removed test for TUS upload.
Will be done in this PR https://github.com/owncloud/ocis/pull/10807

File: tests/acceptance/expected-failures-without-remotephp.md [hunk 13]
```diff
@@ -195,6 +195,9 @@
 - [apiSpaces/publicLink.feature:28](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiSpaces/publicLink.feature#L28)
 - [apiSpaces/publicLink.feature:34](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiSpaces/publicLink.feature#L34)
 - [apiSpaces/publicLink.feature:40](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiSpaces/publicLink.feature#L40)
+- [apiSpaces/uploadSpaces.feature:98](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiSpaces/uploadSpaces.feature#L98)
+- [apiSpaces/uploadSpaces.feature:114](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiSpaces/uploadSpaces.feature#L114)
+- [apiSpaces/uploadSpaces.feature:130](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiSpaces/uploadSpaces.feature#L130)
```
Comment (saw-jan) [inline tests/acceptance/expected-failures-without-remotephp.md:200]: Why do the scenarios fail?

Comment (saw-jan) [reply tests/acceptance/expected-failures-without-remotephp.md:200]: issue without remote.php

**Review by saw-jan — APPROVED** [review]
